File: leetcode/test1851/s2/Test1851.java
package leetcode.test1851.s2;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.PriorityQueue;

public class Test1851 {

    public int[] minInterval(int[][] intervals, int[] queries) {
        var list = new ArrayList<int[]>();
        for (int[] v : intervals) {
            list.add(new int[]{v[0], v[1], -1});
        }
        for (int i = 0;i < queries.length;i++) {
            list.add(new int[]{queries[i],queries[i], i});
        }
        list.sort((o1, o2) -> {
            if (o1[0] == o2[0]) {
                return o1[2] - o2[2];
            }
            return o1[0] - o2[0];
        });
        var pq = new PriorityQueue<int[]>(Comparator.comparingInt(o -> (o[1] - o[0])));
        var res = new int[queries.length];
        for (int[] v : list) {
            int l = v[0], t = v[2];
            if (t == -1) {
                pq.offer(v);
                continue;
            }
            res[t] = -1;
            while (!pq.isEmpty() && pq.peek()[1] < l) {
                pq.poll();
            }
            if (!pq.isEmpty()) {
                res[t] = pq.peek()[1] - pq.peek()[0] + 1;
            }
        }
        return res;
    }
}

File: leetcode/test1851/s2/test1851.go
package s2

import (
	"math"
	"sort"
)

func minInterval(vs [][]int, queries []int) []int {
	sort.Slice(vs, func(i, j int) bool {
		if vs[i][1] != vs[j][1] {
			return vs[i][1] < vs[j][1]
		}
		return vs[i][1]-vs[i][0] > vs[j][1]-vs[j][0]
	})
	var ids = make([]int, len(queries))
	for i := range queries {
		ids[i] = i
	}
	sort.Slice(queries, func(i, j int) bool {
		return queries[ids[i]] < queries[ids[j]]
	})
	l, n := 0, len(vs)
	var res = make([]int, len(queries))
	for _, id := range ids {
		res[id] = math.MaxInt32
		var v = queries[id]
		for l < n && vs[l][1] < v {
			l++
		}
		if l == n || vs[l][0] > v {
			res[id] = -1
			continue
		}
		for ; l < n; l++ {
			res[id] = min(res[id], vs[l][1]-vs[l][0]+1)
			if !(l+1 < n && vs[l+1][0] <= v) {
				break
			}
		}
	}
	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
